File: temas/paradigmasProgramacion/python/objetos.py
"""El paradigma orientado a objetos (POO) organiza el código en clases y objetos, utilizando características como herencia, encapsulamiento y polimorfismo."""


# Definición de la clase base Persona (herencia)
class Persona:
    def __init__(self, nombre, edad):  # Constructor (inicializador de atributos)
        # Encapsulamiento: los atributos son accesibles directamente desde la clase
        self.__nombre = nombre  # El doble guion bajo indica que es un atributo privado (encapsulado)
        self.__edad = edad      # Similar al anterior, el atributo es privado

    # Método público para acceder al nombre
    def get_nombre(self):
        return self.__nombre
    
    # Método público para cambiar el nombre 
    def set_nombre(self, nuevo_nombre):
        self.__nombre = nuevo_nombre

    # Método público para acceder a la edad
    def get_edad(self):
        return self.__edad
    
    # Método público para cambiar la edad 
    def set_edad(self, nueva_edad):
        self.__edad = nueva_edad
    
    # Método de la clase (comportamiento)
    def saludar(self):
        print(f"Hola, mi nombre es {self.__nombre} y tengo {self.__edad} años.")

# Clase Estudiante que hereda de Persona (herencia)
class Estudiante(Persona):
    def __init__(self, nombre, edad, carrera):  # Constructor de la clase hija
        super().__init__(nombre, edad)  # Llamamos al constructor de la clase base (herencia)
        self.__carrera = carrera  # Atributo privado para encapsular la carrera

    # Método público para acceder a la carrera (accesor)
    def get_carrera(self):
        return self.__carrera
    
    # Sobrescribir el método saludar para dar más detalles (polimorfismo)
    def saludar(self):
        # Polimorfismo: la clase hija cambia el comportamiento del método de la clase base
        print(f"Hola, soy {self.get_nombre()}, tengo {self.get_edad()} años y estudio {self.__carrera}.")

# Clase Profesor que también hereda de Persona (herencia)
class Profesor(Persona):
    def __init__(self, nombre, edad, materia):  # Constructor de la clase hija
        super().__init__(nombre, edad)  # Llamamos al constructor de la clase base (herencia)
        self.__materia = materia  # Atributo privado para encapsular la materia que enseña

    # Método público para acceder a la materia (accesor)
    def get_materia(self):
        return self.__materia
    
    # Sobrescribir el método saludar para dar más detalles (polimorfismo)
    def saludar(self):
        # Polimorfismo: cambio de comportamiento en la clase hija
        print(f"Hola, soy el Profesor {self.get_nombre()}, tengo {self.get_edad()} años y enseño {self.__materia}.")

# Crear instancias de las clases
persona1 = Persona("Juan", 30)
estudiante1 = Estudiante("Maria", 22, "Ingeniería")
profesor1 = Profesor("Carlos", 45, "Matemáticas")

# Llamar a los métodos de las clases
persona1.saludar()  # Saludo de una persona
estudiante1.saludar()  # Saludo de un estudiante (polimorfismo en acción)
profesor1.saludar()  # Saludo de un profesor (polimorfismo en acción)

# Cambiar los valores de los atributos privados usando los mutadores
estudiante1.set_nombre("Ana")
estudiante1.set_edad(23)

# Verificar que los valores se actualizaron
estudiante1.saludar()  # Nuevo saludo con el nombre actualizado

File: temas/paradigmasProgramacion/c/objetos.c
/*El paradigma orientado a objetos (OOP) no está directamente soportado en C, ya que C es un lenguaje procedural. Sin embargo, puedes simular algunos aspectos de la programación orientada a objetos utilizando estructuras (`struct`) y punteros para crear algo similar a clases y objetos.

En este ejemplo, crearemos una "clase" en C utilizando estructuras y funciones que operan sobre esas estructuras. Simularemos métodos y atributos para una clase llamada `Persona`, donde se puede establecer y obtener el nombre y la edad de una persona.*/

#include <stdio.h>
#include <stdlib.h> //usado normalmente para administrar memoria dinamicamente
#include <string.h>

// Definición de la "clase" Persona (simulada con struct)
typedef struct {
    char nombre[50];
    int edad;

    // Métodos (funciones que operan sobre la estructura)
    void (*set_nombre)(struct Persona*, const char*); //no devuelve nada, define el puntero a la funcion, y define las variables de las funciones, con sus tipos de datos
    void (*set_edad)(struct Persona*, int);
    void (*mostrar_info)(struct Persona*);
} Persona;

// Función que establece el nombre
void set_nombre(Persona* p, const char* nombre) {
    strncpy(p->nombre, nombre, sizeof(p->nombre) - 1);//"sizeof(p->nombre) - 1" asegura que no quiera copiar un nombre mas grande que el tamanio de la variable "50"
    p->nombre[sizeof(p->nombre) - 1] = '\0'; // Asegurarse de que la cadena esté terminada. determina hasta donde leer
}

// Función que establece la edad
void set_edad(Persona* p, int edad) {
    p->edad = edad;
}

// Función que muestra la información de la persona
void mostrar_info(Persona* p) {
    printf("Nombre: %s\n", p->nombre);
    printf("Edad: %d\n", p->edad);
}

// Función para crear una nueva Persona
Persona* crear_persona(const char* nombre, int edad) {
    Persona* p = (Persona*)malloc(sizeof(Persona));
    if (p == NULL) {
        printf("Error al asignar memoria.\n");
        return NULL;
    }

    // Inicializar los métodos
    p->set_nombre = set_nombre;
    p->set_edad = set_edad;
    p->mostrar_info = mostrar_info;

    // Establecer los valores iniciales
    p->set_nombre(p, nombre);
    p->set_edad(p, edad);

    return p;
}

// Función para liberar la memoria
void destruir_persona(Persona* p) {
    free(p);
}

int main() {
    // Crear una persona usando la "clase"
    Persona* persona1 = crear_persona("Juan", 25);
    
    // Usar los métodos de la "clase"
    persona1->mostrar_info(persona1);

    // Modificar los atributos de la persona
    persona1->set_nombre(persona1, "Carlos");
    persona1->set_edad(persona1, 30);
    
    // Mostrar la nueva información
    persona1->mostrar_info(persona1);

    // Liberar memoria
    destruir_persona(persona1);

    return 0;
}


/* Explicación:
1. **Estructura `Persona`**: Es una simulación de una clase en C. Dentro de la estructura, tenemos dos atributos (`nombre` y `edad`) y tres punteros a funciones, que simulan los métodos de una clase.
   
2. Métodos: Las funciones `set_nombre`, `set_edad` y `mostrar_info` son las que modifican y muestran los atributos de la "clase". Estas funciones se asignan a los punteros de funciones dentro de la estructura `Persona`.

3. `crear_persona`: Esta función crea una nueva instancia de `Persona` (un "objeto" de la clase `Persona`), asignando memoria dinámica para la estructura y configurando sus métodos.

4. Uso en `main`: Creamos una "instancia" de la "clase" `Persona`, usamos sus métodos para modificar sus atributos y mostrar su información. Al final, liberamos la memoria asignada.

¿Qué hemos simulado aquí?
- Clases: Usamos `struct` para simular clases.
- Objetos: Creamos instancias de `Persona` mediante `malloc` (memoria dinámica).
- Métodos: Usamos punteros a funciones dentro de la estructura para simular métodos.
- Encapsulación: Los datos (`nombre` y `edad`) están encapsulados dentro de la estructura `Persona` y sólo pueden ser modificados mediante las funciones asociadas.

Limitaciones de C:
Aunque este enfoque simula la programación orientada a objetos, no tiene algunas características que se encuentran en lenguajes con soporte nativo para OOP, como la herencia, el polimorfismo, y la sobrecarga de métodos. Sin embargo, este enfoque básico puede ser útil para simular OOP en C, especialmente cuando se requiere organización y modularidad en el código.*/